sql/src/main/java/io/crate/planner/operators/WindowAgg.java

```diff
@@ -0,0 +1,151 @@
+import io.crate.planner.PlannerContext;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static io.crate.planner.operators.LogicalPlanner.extractColumns;
+
+public class WindowAgg extends OneInputPlan {
+
+    private final List<WindowFunction> windowFunctions;
+
+    static LogicalPlan.Builder create(LogicalPlan.Builder source, List<WindowFunction> windowFunctions) {
+        if (windowFunctions.isEmpty()) {
+            return source;
+        }
+
+        for (WindowFunction windowFunction : windowFunctions) {
+            WindowDefinition windowDefinition = windowFunction.windowDefinition();
+            if (windowDefinition.partitions().size() > 0 ||
+                windowDefinition.orderBy() != null ||
+                windowDefinition.windowFrameDefinition() != null) {
+                throw new UnsupportedFeatureException("Custom window definitions are currently not supported. " +
+                                                      "Only empty OVER() windows are supported. ");
+            }
+        }
+
+        return (tableStats, usedBeforeNextFetch) -> {
+            HashSet<Symbol> allUsedColumns = new HashSet<>(usedBeforeNextFetch);
+            Set<Symbol> columnsUsedInFunctions = extractColumns(windowFunctions);
+            allUsedColumns.addAll(columnsUsedInFunctions);
+            LogicalPlan sourcePlan = source.build(tableStats, allUsedColumns);
+
+            return new WindowAgg(sourcePlan, windowFunctions);
+        };
+    }
+
+    private WindowAgg(LogicalPlan source, List<WindowFunction> windowFunctions) {
+        super(source);
+        this.windowFunctions = windowFunctions;
+    }
+
+    public List<WindowFunction> windowFunctions() {
+        return windowFunctions;
+    }
+
+    @Override
+    protected LogicalPlan updateSource(LogicalPlan newSource, SymbolMapper mapper) {
+        return new WindowAgg(newSource, windowFunctions);
+    }
+
+    @Override
+    public ExecutionPlan build(PlannerContext plannerContext,
+                               ProjectionBuilder projectionBuilder,
+                               int limit,
+                               int offset,
+                               @Nullable OrderBy order,
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               SubQueryResults subQueryResults) {
+        ExecutionPlan sourcePlan = source.build(
+            plannerContext,
+            projectionBuilder,
+            limit,
+            offset,
+            order,
+            pageSizeHint,
+            params,
+            subQueryResults
+        );
+
+        sourcePlan.addProjection(new WindowAggProjection(groupFunctionsByWindow(source.outputs(), windowFunctions)));
+        return Merge.ensureOnHandler(sourcePlan, plannerContext);
```

Comment:
Shouldn't the `ensureOnHandler` be called before adding the windowAggProjection?